README.md
# trans-server

Transmit files by using **only common Unix/Linux commands, `curl` or `wget`**

| branch | TravisCI | CircleCI | Coverage |
| --- | --- | --- | --- |
| [`master`](https://github.com/nwtgck/trans-server-akka/tree/master) | [![Build Status](https://travis-ci.org/nwtgck/trans-server-akka.svg?branch=master)](https://travis-ci.org/nwtgck/trans-server-akka) | [![CircleCI](https://circleci.com/gh/nwtgck/trans-server-akka/tree/master.svg?style=shield)](https://circleci.com/gh/nwtgck/trans-server-akka/tree/master) | [![Coverage Status](https://coveralls.io/repos/github/nwtgck/trans-server-akka/badge.svg?branch=master)](https://coveralls.io/github/nwtgck/trans-server-akka?branch=master) |
| [`develop`](https://github.com/nwtgck/trans-server-akka/tree/develop) | [![Build Status](https://travis-ci.org/nwtgck/trans-server-akka.svg?branch=develop)](https://travis-ci.org/nwtgck/trans-server-akka) | [![CircleCI](https://circleci.com/gh/nwtgck/trans-server-akka/tree/develop.svg?style=shield)](https://circleci.com/gh/nwtgck/trans-server-akka/tree/develop) | [![Coverage Status](https://coveralls.io/repos/github/nwtgck/trans-server-akka/badge.svg?branch=develop)](https://coveralls.io/github/nwtgck/trans-server-akka?branch=develop) |

## Send a file by wget

![wget_send](demos/wget_send.gif)

## Download the file by wget

![wget_get](demos/wget_get.gif)

## Public Server on Heroku

https://trans-akka.herokuapp.com/


## Quick Start

You can send, download and delete a file by the following commands.

```
# Send a file
wget -q -O - https://trans-akka.herokuapp.com/ --post-file=./test.txt
```

```
# Downlaod the file
wget https://trans-akka.herokuapp.com/a7h
# ('a7h' is a File ID created after sending)
```

```
# Delete a file
wget -q -O - --method DELETE https://trans-akka.herokuapp.com/a7h
```


### Option explanations

* `-q`: No prgress bar
* `-O -`: Response to stdout


## Main features

* Send/Get by **only common commands**, `curl` or `wget`
* Send/Get **on your browser**
* Send/Get with restrictions
  - download key
  - store duration
  - how many times you can download
* Delete files you sent
  - without delete key
  - with delete key
* Change File ID length (for security)
* Encrypt and Compress file for security

## How to run the server

You can choose any ways you want bellow.

### Way 1 - Run on Docker

[![Docker Automated build](https://img.shields.io/docker/automated/nwtgck/trans-server-akka.svg)](https://hub.docker.com/r/nwtgck/trans-server-akka/)

```bash
docker run -p 8080:8181 nwtgck/trans-server-akka
```

Then you can go http://localhost:8080/

#### Docker run for daemonize and data persistence

You can also run the following command for **daemonize** and **data persistence**.

```bash
docker run -d -p 8080:8181 -v $PWD/trans-db:/trans/db --restart=always nwtgck/trans-server-akka
```

Data will be stored in `$PWD/trans-db` on your host machine. (Currently file-base H2 database is used, and files sent are stored as compressed and encrypted files)

### Way 2 - sbt "run-main ..."

The following example runs a server on 80 port.
```sh
$ cd <this-project>
$ ./make-keystore.sh
$ sbt "run-main io.github.nwtgck.trans_server.Main --http-port=8282"
```

### Way 3 - Making a jar

#### 1. Make a keystore

```sh
$ cd <this-project>
$ ./make-keystore.sh
```


#### 2. Make a jar

```sh
$ cd <this-project>
$ sbt assembly
```
(It takes a time)

#### 3. Run the jar

```sh
$ java -jar target/scala-2.11/trans-server-akka.jar --http-port=8282 --https-port=4433
```

## How to send a file to the server

### Way 1 - curl

The following example is sending `../test.txt`

```
$ curl https://trans-akka.herokuapp.com/ --data-binary @../test.txt
```

#### output
```
ab2
```

The server response, `ab2` is a File ID to get `../test.txt`


### Way 2 - wget

The following example sends `../test.txt`

```sh
$ wget -q -O - https://trans-akka.herokuapp.com/ --post-file=../test.txt
```

* `-q` is for non-progress bar
* `-O -` is to output STDOUT  

#### output
```
9vi
```

The server response, `9vi` is a File ID to get `../test.txt`


### Way 3 - wc & cat & nc

**This way is for a user which can't use `curl` command.**

The following example sends `sounds.zip`.

#### 1. Get a file byte

```sh
$ wc -c  < sounds.zip
```


##### output
```
1161257298 # Use it later
```

#### 2. Create `header.txt`

```
POST / HTTP/1.1
Host: localhost:4343
Content-Length: 1161257298

```

Don't forget the end of `'\n'`

 #### 3. Send a file by HTTP request


```
$ cat header.txt sounds.zip | nc localhost 4343
```


 The response is bellow.
```
HTTP/1.1 200 OK
Server: akka-http/10.0.5
Date: Sun, 23 Apr 2017 04:22:32 GMT
Connection: close
Content-Type: text/plain; charset=UTF-8
Content-Length: 4

6oz
```

`6oz` is FILE ID.


## How to get a file from the server

### Way 1 - wget

```sh
$ wget https://trans-akka.herokuapp.com/ab2
```

`ab2` is a File ID.

### Way 2 - curl


```sh
$ curl https://trans-akka.herokuapp.com/ab2 > test.txt
```

`ab2` is a File ID.


### Way 3 - Using a Browser

Open `https://trans-akka.herokuapp.com/ab2` on your browser

`ab2` is a File ID.


## Sending-options

Here is options you can use when sending

|GET parameter | default value | decription |
|---|---:|---|
| `duration`    | 1 hour        | Store duration/life                                  |
| `get-times`   | infinite      | How many times you can download                      |
| `id-length`   | `3`           | Length of File ID                                    |
| `deletable`   | `true`        | Whether a file can be deleted                        |
| `delete-key`  | nothing       | Key for deletion                                     | 
| `secure-char` | `false`       | Whether more complex characters are used for File ID | 

### An example with options

```bash
wget -q -O - 'https://trans-akka.herokuapp.com/?duration=30s&get-times=1&id-length=16&delete-key=mykey1234&secure-char' --post-file=./hello.txt
```

The command means
* duration is 30 seconds
* download once
* File ID length is 16
* The file is deletable and key is `'mykey1234'`
* Use more complex characters for File ID

### Available duration examples
* `10s` - 10 seconds
* `2m`  - 2 minutes
* `12h` - 12 hours
* `25d` - 25 days

### Usage of `deletable`

All bellow are valid usages.

```bash
'https://trans-akka.herokuapp.com/?deletable'
# (same meaning as `deletable=true`)
```

```bash
'https://trans-akka.herokuapp.com/?deletable=true'
```

```bash
'https://trans-akka.herokuapp.com/?deletable=false'
```

### Set download key

```bash
curl -u dummy_user:mypass123 -T ./hello.txt  https://trans-akka.herokuapp.com
```

When you don't want to type key in command, you can use `-u dummy_user` instead of `-u dummy_user:mypass123`. Then, `curl` provide you a password prompt.

NOTE: `dummy_user` can be change any name including empty `''`.


### Get with download key

```bash
curl -u dummy_user:mypass123 https://trans-akka.herokuapp.com/yk2
```

NOTE: The server doesn't check user name. So, you can use any name for user name, not only `"dummy_user"`.


## Delete file

## delete without delete key

```bash
# wget version
wget -q -O - --method=DELETE 'https://trans-akka.herokuapp.com/vua'
```

```bash
# curl version
curl -X DELETE 'https://trans-akka.herokuapp.com/vua'
```
(`vua` is a File ID)

## delete with delete key

```bash
# wget version
wget -q -O - --method=DELETE 'https://trans-akka.herokuapp.com/6em?delete-key=1234'
```

```bash
# curl version
curl -X DELETE 'https://trans-akka.herokuapp.com/6em?delete-key=1234'
```

## My Note

[MYNOTE.md](MYNOTE.md)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)

## [Unreleased]

## [1.19.8] - 2019-03-14
### Changed
- Update h2 database to 1.4.199
- Update dependencies

## [1.19.7] - 2019-02-23
### Changed
- Update h2 database to 1.4.198

## [1.19.6] - 2019-01-12
### Changed
- Update dependencies

## [1.19.5] - 2019-01-08
### Changed
- Use openjdk:8-alpine as base Docker image

## [1.19.4] - 2019-01-06
### Changed
- Update dependencies
- Use Alpine-based docker image

## [1.19.3] - 2018-11-26
### Added
- Support Glitch

## [1.19.2] - 2018-11-13
### Fixed
- Fix to delete the file if store process is failed

## [1.19.1] - 2018-11-12
### Fixed
- Fix to notify an error to user when store process is failed

## [1.19.0] - 2018-11-10
### Added
- Allow users to redirect URI by `/r/<File ID>`

## [1.18.0] - 2018-09-05
### Added
- Allow users to specify File ID when sending

## [1.17.0] - 2018-07-31
### Added
- Add `--db-url` to specify database URL like 'jdbc:h2:tcp://localhost/~/h2db/trans'
- Allow users to send data via GET method

### Changed
- Allow users to send bigger data by HTTP GET method

## [1.16.2] - 2018-07-18
### Changed
- Improve `/help` to use simpler commands and add CLI installation written in Go

### Fixed
- Fix error HTTP statuses

## [1.16.0] - 2018-07-16
### Added
- Redirect HTTP to HTTPS when "X-Forwarded-Proto" is used (Heroku or IBM Cloud (Bluemix))
- Redirect HTTP to HTTPS in top page by specifying a command line option

### Changed
- Encrypt storing files by download-key from Basic Authentication
- Specify HTTP/HTTPS ports by command line options

## [1.15.1] - 2018-07-13
### Added
- Show file ID history on Web client

## [1.15.0] - 2018-07-12
### Added
- Allow users to set download key

## [1.14.0] - 2018-07-09
### Added
- Add `X-FILE-MD5` HTTP Header for checksum
- Add `X-FILE-SHA1` HTTP Header for checksum
- Add `X-FILE-SHA256` HTTP Header for checksum

## [1.13.0] - 2018-06-30
### Added
- Add `secure-char` GET parameter for more complex File ID

## [1.12.0] - 2018-06-28
### Added
- Adapt to deletion in Web client

### Changed
- Allow cross origin access

## [1.11.0] - 2018-06-24
### Added
- Adapt to sending a file by PUT method

### Fixed
- Change maximum file ID length to 128 for file system limitation

## [1.10.0] - 2018-06-24
### Changed
- Allow user to get file by any end-path

## [1.9.0] - 2018-05-22
### Changed
- Not to generate file ID including misunderstandable chararacters

## [1.8.0] - 2018-04-06
### Added
- Add text-sending form in web client

## [1.7.1] - 2018-04-01
### Fixed
- Fix not to fail to upload big file via multipart because of size limit

## [1.7.0] - 2018-01-22
### Added
- Add [manifest.yml](manifest.yml) for [IBM Cloud](https://www.ibm.com/cloud/)
### Changed
- Make `/help` page to be easy to copy

## [1.6.0] - 2017-12-03
### Added
- GET `/help` routing for help
- GET `/version` routing for version

### Fixed
- Return "Content-Length" in HTTP response to adapt [`pget`](https://github.com/Code-Hex/pget)

## [1.5.1] - 2017-12-01
### Changed
- Ignore cases, for example `true`, `True` or `TRUE` in `deletable` parameter
- Allow to users to specify keystore outside of Docker container

## [1.5.0] - 2017-11-27
### Changed
- Rename GET parameters
  - `key` => `delete-key`
  - `length` => `id-length`
  - `times` => `get-times` 

## [1.4.0] - 2017-11-27
### Changed
- Encrypt stored files in the server
- Compress stored files in the server

## [1.3.1] - 2017-11-26
### Added
- Specify options in the Web client

## [1.3.0] - 2017-11-25
## Added
- Allow users to change length of File ID
- Allow users to delete files with key or without key

## [1.2.0] - 2017-11-25
### Added
- Allow users to specify download-times limit


## [1.1.1] - 2017-11-25
### Fixed
- Docker automated problem with `sbt-assembly` 


## [1.1.0] - 2017-11-25
### Added
- Allow users to limit file by store duration


## 1.0.0 - 2017-11-23
### Added
- Basic sending file and getting file via HTTP
- HTTPS connection by using keystore
- Make jar by [`sbt-assembly`](https://github.com/sbt/sbt-assembly)
- Adapt [`sbt-native-packager`](https://github.com/sbt/sbt-native-packager) for [Heroku](https://heroku.com/)


[Unreleased]: https://github.com/nwtgck/trans-server-akka/compare/v1.19.8...HEAD
[1.19.8]: https://github.com/nwtgck/trans-server-akka/compare/v1.19.7...v1.19.8
[1.19.7]: https://github.com/nwtgck/trans-server-akka/compare/v1.19.6...v1.19.7
[1.19.6]: https://github.com/nwtgck/trans-server-akka/compare/v1.19.5...v1.19.6
[1.19.5]: https://github.com/nwtgck/trans-server-akka/compare/v1.19.4...v1.19.5
[1.19.4]: https://github.com/nwtgck/trans-server-akka/compare/v1.19.3...v1.19.4
[1.19.3]: https://github.com/nwtgck/trans-server-akka/compare/v1.19.2...v1.19.3
[1.19.2]: https://github.com/nwtgck/trans-server-akka/compare/v1.19.1...v1.19.2
[1.19.1]: https://github.com/nwtgck/trans-server-akka/compare/v1.19.0...v1.19.1
[1.19.0]: https://github.com/nwtgck/trans-server-akka/compare/v1.18.0...v1.19.0
[1.18.0]: https://github.com/nwtgck/trans-server-akka/compare/v1.17.0...v1.18.0
[1.17.0]: https://github.com/nwtgck/trans-server-akka/compare/v1.16.2...v1.17.0
[1.16.2]: https://github.com/nwtgck/trans-server-akka/compare/v1.16.1...v1.16.2
[1.16.0]: https://github.com/nwtgck/trans-server-akka/compare/v1.15.1...v1.16.0
[1.15.1]: https://github.com/nwtgck/trans-server-akka/compare/v1.15.0...v1.15.1
[1.15.0]: https://github.com/nwtgck/trans-server-akka/compare/v1.14.0...v1.15.0
[1.14.0]: https://github.com/nwtgck/trans-server-akka/compare/v1.13.0...v1.14.0
[1.13.0]: https://github.com/nwtgck/trans-server-akka/compare/v1.12.0...v1.13.0
[1.12.0]: https://github.com/nwtgck/trans-server-akka/compare/v1.11.0...v1.12.0
[1.11.0]: https://github.com/nwtgck/trans-server-akka/compare/v1.10.0...v1.11.0
[1.10.0]: https://github.com/nwtgck/trans-server-akka/compare/v1.9.0...v1.10.0

src/main/scala/Setting.scala
import scala.concurrent.duration.{DurationInt, FiniteDuration}

/**
  * Created by Ryo on 2017/04/23.
  */
object Setting {
  val HOST                 : String         = "0.0.0.0"
  val DEFAULT_HTTP_PORT    : Int            = 8181
  val DEFAULT_HTTPS_PORT   : Int            = 4343
  val File_DB_PATH         : String         = "db/file_db"
  val KEY_STORE_PATH       : String         = "trans.keystore"
  val KEY_STORE_PASSWORD   : String         = "changeit"
  val DefaultStoreDuration : FiniteDuration = 1.hour
  val MaxStoreDuration     : FiniteDuration = 30.days
  val CleanupDuration      : FiniteDuration = 1.minute
  val DefaultIdLength      : Int            = 3
  val MinIdLength          : Int            = 3
  val MaxIdLength          : Int            = 256
  val KeySalt              : String         = "ymiKicjOq5M3yIMmVBEPFcLnkNxJ4n2iY9ms3fYRsBDS3wZvS1lev9ToLnZhlj3O"
  val FileEncryptionKey    : String         = "vKOhINCMuO47xxel" // NOTE: Its length should be 16
  val FileIdGenTryLimit    : Int            = 500 // NOTE: Simple study shows 500 is enough (https://github.com/nwtgck/random-str-collision-graph-scala)

  // Reserved route name of GET request
  object GetRouteName{
    val Multipart: String = "multipart"
    val Help     : String = "help"
    val Version  : String = "version"

    // Reserved route names
    val allSet   : Set[String] = Set(
      Multipart,
      Help,
      Version
    )
  }
}
